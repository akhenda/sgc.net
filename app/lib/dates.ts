import dayjs from 'dayjs';
import customParseFormat from 'dayjs/plugin/customParseFormat';

import type { Sighting, Sightings } from './api/resources/ufos';

dayjs.extend(customParseFormat);

var days = ['Sunday', 'Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday'];

export function convertDateFormat(dateString: string) {
  // Use dayjs to parse the date string with the specified format
  const dateObject = dayjs(dateString, 'DD/MM/YYYY');

  // Handle cases where the date string is not valid
  if (!dateObject.isValid()) throw new Error('Invalid Date');

  return dateObject.toDate();
}

export function getISOWeek(date: Date): number {
  const d = new Date(date.getTime());

  d.setDate(d.getDate() + 4 - (d.getDay() || 7));

  const yearStart = new Date(d.getFullYear(), 0, 1);
  const weekNo = Math.ceil(((d.getTime() - yearStart.getTime()) / 86400000 + 1) / 7);

  return weekNo;
}

type ChartSighting = Sighting & { name: string; prettyDate: string; week: string };
type ChartSightings = ChartSighting[];

export interface WeekData {
  name: string;
  weekStartDate: string; // store the start date of the week
  sightings: ChartSightings; // array of daily sightings within this week
}

export interface WeeklySightings {
  [weekKey: string]: WeekData; // keys will be like "YYYY WNN"
}

export function groupDataByWeek(data: Sightings): WeeklySightings {
  const weeklyData: WeeklySightings = {};

  data.forEach((entry) => {
    const date = convertDateFormat(entry.date);

    const dayOfWeek = date.getDay();
    const daysToSubtract = dayOfWeek === 0 ? 6 : dayOfWeek - 2;
    const weekStartDate = new Date(date);

    weekStartDate.setDate(date.getDate() - daysToSubtract);

    const weekStartDateString = weekStartDate.toISOString().split('T')[0];
    const weekYear = weekStartDate.getFullYear();
    const weekNumber = getISOWeek(date); // Calculate week based on the original date
    const weekKey = `${weekYear} W${weekNumber.toString().padStart(2, '0')}`;

    if (!weeklyData[weekKey]) {
      weeklyData[weekKey] = { weekStartDate: weekStartDateString, sightings: [], name: weekKey };
    }

    weeklyData[weekKey].sightings.push({
      ...entry,
      name: days[dayOfWeek],
      prettyDate: date.toDateString(),
      week: weekKey,
    });
  });

  // Fill in missing dates for each week
  Object.entries(weeklyData).forEach(([_, weekData]) => {
    const weekStartDate = new Date(weekData.weekStartDate);
    const datesPresent = new Set(weekData.sightings.map((s) => s.date));
    const filledSightings: ChartSightings = [];

    for (let i = 0; i < 7; i++) {
      const currentDate = new Date(weekStartDate);

      currentDate.setDate(weekStartDate.getDate() + i);

      const currentDateString = dayjs(currentDate).format('DD/MM/YYYY');

      if (datesPresent.has(currentDateString)) {
        const existingEntry = weekData.sightings.find((s) => s.date === currentDateString);
        if (existingEntry) filledSightings.push(existingEntry);
      } else {
        filledSightings.push({
          date: currentDateString,
          sightings: 0,
          name: days[currentDate.getDay()],
          prettyDate: currentDate.toDateString(),
          week: weekData.name,
        });
      }
    }

    // Sort the sightings by date to ensure chronological order
    filledSightings.sort((a, b) => {
      const [dayA, monthA, yearA] = a.date.split('/').map(Number);
      const dateA = new Date(yearA, monthA - 1, dayA);
      const [dayB, monthB, yearB] = b.date.split('/').map(Number);
      const dateB = new Date(yearB, monthB - 1, dayB);
      return dateA.getTime() - dateB.getTime();
    });

    weekData.sightings = filledSightings;
  });

  return weeklyData;
}
