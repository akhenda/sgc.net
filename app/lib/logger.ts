import type { ReactNode } from 'react';

import { consola, type LogType } from 'consola';
import { type ExternalToast, toast as sonnerToast } from 'sonner';

import type { AnyValue } from '../types';

export type LogLevel = LogType;

// Type for the object returned by log methods, enabling chaining
interface LogResult {
  toast: (message?: string | ReactNode, options?: ExternalToast) => void;
}

// Error Reporter Helper
function reportErrorToSentry(
  _level: LogLevel,
  _message: string,
  _optionalParams: AnyValue[],
): void {
  // report to sentry or maybe log rocket
}

function logInternal(level: LogLevel, message: string, optionalParams: AnyValue[]): void {
  const logFn = consola[level] || console.log;

  try {
    logFn(message, ...optionalParams);
  } catch (logError) {
    console.error('Custom logger failed:', logError);
    console.error('Original log:', { level, message, optionalParams });
  }

  reportErrorToSentry(level, message, optionalParams);
}

function createLogMethod(level: LogLevel) {
  return (message: string, ...optionalParams: AnyValue[]): LogResult => {
    // Perform the core logging action
    logInternal(level, message, optionalParams);

    // Return object for chaining, conditionally adding .toast
    const result: LogResult = { toast: () => null };

    result.toast = (toastMessage?: string | ReactNode, options?: ExternalToast): void => {
      try {
        // Use the provided toast message, or fallback to the original log message
        const displayMessage = toastMessage ?? message;

        // Map log level to sonner toast type (optional, customize as needed)
        switch (level) {
          case 'error':
          case 'fatal':
            sonnerToast.error(displayMessage, options);
            break;
          case 'warn':
            sonnerToast.warning(displayMessage, options);
            break;
          case 'info':
            sonnerToast.info(displayMessage, options);
            break;
          case 'success':
            sonnerToast.success(displayMessage, options);
            break;
          default:
            sonnerToast(displayMessage, options); // Default toast
        }
      } catch (toastError) {
        consola.error('Failed to show toast:', toastError);
      }
    };

    return result;
  };
}

export const logger = {
  trace: createLogMethod('trace'),
  debug: createLogMethod('debug'),
  success: createLogMethod('success'),
  info: createLogMethod('info'),
  warn: createLogMethod('warn'),
  error: createLogMethod('error'),
  fatal: createLogMethod('fatal'),
};
