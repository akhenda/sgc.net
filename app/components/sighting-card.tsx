import type { LucideIcon } from 'lucide-react';

import type { Sighting } from '@/lib/api/resources/ufos';
import { convertDateFormat } from '@/lib/dates';

type SightingCardProps = { title: string; sighting?: Sighting | null; Icon: LucideIcon };

export function SightingCard({ title, sighting, Icon }: SightingCardProps) {
  if (!sighting) return null;

  const { date, sightings } = sighting;

  return (
    <div className="relative flex min-h-40 flex-1 flex-col justify-between overflow-hidden rounded-md bg-card p-6 shadow-lg">
      <h3 className="font-bold text-sm">{title}</h3>
      {date && <p className="text-muted-foreground">{convertDateFormat(date).toDateString()}</p>}
      {(sightings || sightings === 0) && (
        <p className="text-4xl text-muted-foreground">{sightings}</p>
      )}
      <Icon className="absolute right-4 bottom-4 h-22 w-22 opacity-5" />
    </div>
  );
}
