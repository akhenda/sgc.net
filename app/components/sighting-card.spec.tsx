import { render, screen } from '@testing-library/react';
import type { LucideIcon } from 'lucide-react';
import { describe, expect, it, vi } from 'vitest';

import type { Sighting } from '../lib/api/resources/ufos';
import { convertDateFormat } from '../lib/dates';

import { SightingCard } from './sighting-card';

const MockIcon = vi.fn(({ className }) => (
  <svg data-testid="mock-icon" className={className} />
)) as unknown as LucideIcon;

vi.mock('../lib/dates', () => ({
  convertDateFormat: vi.fn((dateString: string) => {
    const [day, month, year] = dateString.split('/').map(Number);

    return new Date(year, month - 1, day);
  }),
}));

describe('SightingCard', () => {
  const mockTitle = 'Total Sightings';
  const mockSighting: Sighting = { date: '15/05/2023', sightings: 42 };

  it('should render the card with sighting data', () => {
    render(<SightingCard title={mockTitle} sighting={mockSighting} Icon={MockIcon} />);

    const expectedDateString = new Date(2023, 4, 15).toDateString(); // May 15, 2023

    expect(screen.getByText(mockTitle)).toBeInTheDocument();
    expect(screen.getByText(expectedDateString)).toBeInTheDocument();
    expect(screen.getByText(mockSighting.sightings.toString())).toBeInTheDocument();
    expect(screen.getByTestId('mock-icon')).toBeInTheDocument();
    expect(MockIcon).toHaveBeenCalled();
  });

  it('should return null if no sighting data is provided', () => {
    const { container } = render(
      <SightingCard title={mockTitle} sighting={null} Icon={MockIcon} />,
    );

    expect(container.firstChild).toBeNull();
  });

  it('should render with 0 sightings', () => {
    const sightingWithZero: Sighting = { date: '01/01/2024', sightings: 0 };

    render(<SightingCard title={mockTitle} sighting={sightingWithZero} Icon={MockIcon} />);

    expect(screen.getByText('0')).toBeInTheDocument();
  });

  it('should call convertDateFormat with the correct date string', () => {
    render(<SightingCard title={mockTitle} sighting={mockSighting} Icon={MockIcon} />);

    expect(convertDateFormat).toHaveBeenCalledWith(mockSighting.date);
  });

  it('should match snapshot with sighting data', () => {
    const { container } = render(
      <SightingCard title={mockTitle} sighting={mockSighting} Icon={MockIcon} />,
    );

    expect(container).toMatchSnapshot();
  });
});
