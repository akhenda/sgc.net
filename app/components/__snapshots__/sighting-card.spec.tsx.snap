// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SightingCard > should match snapshot with sighting data 1`] = `
<div>
  <div
    class="relative flex min-h-40 flex-1 flex-col justify-between overflow-hidden rounded-[1rem] bg-card p-6 shadow-lg"
  >
    <h3
      class="font-bold text-sm"
    >
      Total Sightings
    </h3>
    <p
      class="text-muted-foreground"
    >
      Mon May 15 2023
    </p>
    <p
      class="text-4xl text-muted-foreground"
    >
      42
    </p>
    <svg
      class="absolute right-4 bottom-4 h-22 w-22 opacity-5"
      data-testid="mock-icon"
    />
  </div>
</div>
`;
